src/darwin/Framework/CHIP/MTROTAImageTransferHandler.mm

```diff
@@ -283,6 +304,30 @@ - (instancetype)initWithMTROTAImageTransferHandler:(MTROTAImageTransferHandler *
         return CHIP_ERROR_INCORRECT_STATE;
     }
 
+    void (^completionHandler)(NSData * _Nullable data, BOOL isEOF) = nil;
+
+    // If the peer node is a Thread device, check how much time has elapsed since we started processing the BlockQuery.
+    // If the time elapsed is greater than kBdxThrottleDefaultIntervalInMsecs, call the completion handler to respond with a Block right away.
+    // If time elapsed is less than kBdxThrottleDefaultIntervalInMsecs, dispatch the completion handler to respond with a Block after kBdxThrottleDefaultIntervalInMsecs has elapsed.
+
+    if (mIsPeerNodeAKnownThreadDevice) {
+        completionHandler = ^(NSData * _Nullable data, BOOL isEOF) {
+            auto timeElapsed = System::SystemClock().GetMonotonicTimestamp() - startBlockQueryHandlingTimestamp;
+
+            if (timeElapsed >= mBDXThrottleIntervalForThreadDevicesInMSecs) {
+                respondWithBlock(data, isEOF);
+            } else {
+                auto timeRemaining = std::chrono::duration_cast<std::chrono::nanoseconds>(mBDXThrottleIntervalForThreadDevicesInMSecs - timeElapsed);
+                dispatch_time_t time = dispatch_time(DISPATCH_TIME_NOW, (int64_t) (timeRemaining.count()));
```

Comment:
The cast should not be needed, but if it's needed it should be a `static_cast`, not an unsafe C-style cast.

Reply:
Addressed in https://github.com/project-chip/connectedhomeip/pull/37562